dll/ntdll/ldr/verifier.c

```diff
@@ -0,0 +1,685 @@
+                Provider->ProviderDlls = Descriptor->ProviderDlls;
+                Provider->ProviderDllLoadCallback = Descriptor->ProviderDllLoadCallback;
+                Provider->ProviderDllUnloadCallback = Descriptor->ProviderDllUnloadCallback;
+                Provider->ProviderNtdllHeapFreeCallback = Descriptor->ProviderNtdllHeapFreeCallback;
+
+                /* Update some info for the provider */
+                Descriptor->VerifierImage = LdrpImageEntry->BaseDllName.Buffer;
+                Descriptor->VerifierFlags = AVrfpVerifierFlags;
+                Descriptor->VerifierDebug = AVrfpDebug;
+
+                /* We don't have these yet */
+                DPRINT1("AVRF: RtlpGetStackTraceAddress MISSING\n");
+                DPRINT1("AVRF: RtlpDebugPageHeapCreate MISSING\n");
+                DPRINT1("AVRF: RtlpDebugPageHeapDestroy MISSING\n");
+                Descriptor->RtlpGetStackTraceAddress = NULL;
+                Descriptor->RtlpDebugPageHeapCreate = NULL;
+                Descriptor->RtlpDebugPageHeapDestroy = NULL;
+                Status = STATUS_SUCCESS;
+            }
+            else
+            {
+                DbgPrint("AVRF: provider %wZ passed an invalid descriptor @ %p\n", &Provider->DllName, Descriptor);
+                Status = STATUS_INVALID_PARAMETER_4;
+            }
+        }
+        else
+        {
+            DbgPrint("AVRF: provider %wZ did not initialize correctly\n", &Provider->DllName);
+            Status = STATUS_DLL_INIT_FAILED;
+        }
+    }
+    _SEH2_EXCEPT(EXCEPTION_EXECUTE_HANDLER)
+    {
+        Status = _SEH2_GetExceptionCode();
+    }
+    _SEH2_END;
+
+    if (!NT_SUCCESS(Status))
+        return Status;
+
+
+    if (AVrfpDebug & RTL_VRF_DBG_LISTPROVIDERS)
+        DbgPrint("AVRF: initialized provider %wZ (descriptor @ %p)\n", &Provider->DllName, Descriptor);
+
+    /* Done loading providers, allow dll notifications */
+    AVrfpInitialized = TRUE;
+
+    AVrfpChainDuplicateThunks();
+    AVrfpResnapInitialModules();
+
+    /* Manually call with DLL_PROCESS_ATTACH, since the process is not done initializing */
+    _SEH2_TRY
+    {
+        if (!LdrpCallInitRoutine(Provider->EntryPoint,
+                                 Provider->BaseAddress,
+                                 DLL_PROCESS_ATTACH,
+                                 NULL))
+        {
+            DbgPrint("AVRF: provider %wZ did not initialize correctly\n", &Provider->DllName);
+            Status = STATUS_DLL_INIT_FAILED;
+        }
+
+    }
+    _SEH2_EXCEPT(EXCEPTION_EXECUTE_HANDLER)
+    {
+        Status = _SEH2_GetExceptionCode();
+    }
+    _SEH2_END;
+
+    return Status;
+}
+
+
+NTSTATUS
+NTAPI
+AVrfInitializeVerifier(VOID)
+{
+    NTSTATUS Status;
+    PVERIFIER_PROVIDER Provider;
+    PLIST_ENTRY Entry;
+    WCHAR* Ptr, *Next;
+
+    Status = RtlInitializeCriticalSection(&AVrfpVerifierLock);
+    InitializeListHead(&AVrfpVerifierProvidersList);
+
+    if (!SUCCEEDED(Status))
```

Comment:
NT_SUCCESS

Reply:
Ha shit I did it again.

Thanks for reviewing :)